text/0047-assignment-of-availability-chunks.md

```diff
@@ -0,0 +1,313 @@
+
+> A validator should request chunks by picking peers randomly and must recover at least `f+1` chunks, where
+`n=3f+k` and `k in {1,2,3}`.
+
+For parity's polkadot node implementation, the process was further optimised. At this moment, it works differently based
+on the estimated size of the available data:
+
+(a) for small PoVs (up to 128 Kib), sequentially try requesting the unencoded data from the backing group, in a random
+order. If this fails, fallback to option (b).
+
+(b) for large PoVs (over 128 Kib), launch N parallel requests for the erasure coded chunks (currently, N has an upper
+limit of 50), until enough chunks were recovered. Validators are tried in a random order. Then, reconstruct the
+original data.
+
+All options require that after reconstruction, validators then re-encode the data and re-create the erasure chunks trie
+in order to check the erasure root.
+
+### Availability recovery from systematic chunks
+
+As part of the effort of
+[increasing polkadot's resource efficiency, scalability and performance](https://github.com/paritytech/roadmap/issues/26),
+work is under way to modify the Availability Recovery protocol by leveraging systematic chunks. See
+[this comment](https://github.com/paritytech/polkadot-sdk/issues/598#issuecomment-1792007099) for preliminary
+performance results.
+
+In this scheme, the relay chain node will first attempt to retrieve the ~N/3 systematic chunks from the validators that
+should hold them, before falling back to recovering from regular chunks, as before.
+
+A re-encoding step is still needed for verifying the erasure root, so the erasure coding overhead cannot be completely
+brought down to 0.
+
+### Chunk assignment function
+
+#### Properties
+
+The function that decides the chunk index for a validator should be parameterized by at least
+`(validator_index, relay_parent, core_index)`
+and have the following properties:
+1. deterministic
+1. relatively quick to compute and resource-efficient.
+1. when considering the other params besides `validator_index` as fixed, the function should describe a permutation
+of the chunk indices
+1. considering `relay_parent` as a fixed argument, the validators that map to the first N/3 chunk indices should
+have as little overlap as possible for different paras scheduled on that relay parent.
+
+In other words, we want a uniformly distributed, deterministic mapping from `ValidatorIndex` to `ChunkIndex` per block
+per core.
+
+It's desirable to not embed this function in the runtime, for performance and complexity reasons.
+However, this means that the function needs to be kept very simple and with minimal or no external dependencies.
+Any change to this function could result in parachains being stalled and needs to be coordinated via a runtime upgrade
+or governance call.
+
+#### Proposed function
+
+Pseudocode:
+
+```rust
+pub fn get_chunk_index(
+  n_validators: u32,
+  validator_index: ValidatorIndex,
+  block_number: BlockNumber,
```

Comment:
Ok using the relay parent of the candidate is at the very least not ideal because with async backing, the used relay parent could vary from candidate to candidate in the same block, which means that the equal load distribution might end up not being that equal after all.

Reply:
@burdges WDYT of this?

Reply:
I'd expect the opposite:  After rewards exist, you choose your core so that friendly validators have more systemic chunk assignments.  It annoys some validators who become slightly less profitable.   It annoys polkadot because then unprepared validators could recieve more systemic chunk assignments than they could handle.  It's probably fine since those unprepared validatoprs have the backers as systemic chunks backup, not just themselves.

A simple rotation suffices here imho since any adversary must specifically code the rotation into their core purchasing, so anyone not doing this gets into their way, but they can rotate very coarsely I guess.

If we do not rotate then some cores could naturally do this, which could prompt some stupid behavior within the community.  Ain't too big a deal, but maybe easier to rotate by slot, unless we've some technical reason otherwise.

Reply:
> Ain't too big a deal since again 1/3 gets pretty coarse, but maybe easier to rotate by slot, unless we've some technical reason otherwise.

rotating by block number/slot will require that the chunk request requester has access to the relay parent of the candidate if they want to perform systematic recovery.
Considering that core index will be added to the candidate receipt, if we were to only use the core index for the rotation, we would only need the `CandidateReceipt` in order to perform systematic recovery or get our assigned chunk. Which would be easier and would also enable systematic recovery for all dispute participants that may not have access to the relay chain blocks for remote disputes.

Reply:
I would really appreciate if all information could be self contained. @burdges I don't get your argument. Systemic chunks will be distributed fairly equally among cores in a block. We don't need rotation for this: First core, first x validators are systemic chunk providers, next core , next x validators, .... 

The only difference is, if we assume that some cores are predictably higher loaded than others (PoVs larger in expectation). This can happen, but could be mitigated by incorporating in rewards: You reward people in proportion to the size of the chunks they served. Should be possible to design this in a way, so there is no meaningful advantage in trying to manipulate for which cores your validators (or others) are responsible for.

In general we should strive for (and agile coretime is one important step into that direction), that cores are fully loaded as much as possible: Better allocation efficiency. If we achieve that goal, then having no rotation should not matter at all anymore.

Also in case this ever proves to be a real problem still, we could adjust the scheduler to move parachains between cores all the time, which would then still give us self contained systemic recovery, with evened out load. I don't think we we will ever need that, just pointing out that we could achieve rotation via different means, which don't affect availability recovery at all.

There exists some uneven load, even if all cores are equal, but:

With 100 cores, two thirds of the validators will be systemic chunk providers for 33 cores. One third of them will be systemic chunk providers for 34 cores. That is as bad as it gets, when it comes to unequal load in number of cores directly. It also has the good property that this difference gets smaller with more cores. We can also even it out, by always increasing core count in steps of three, if we wanted to get rid of this entirely.

Reply:
Isn't the relay parent given in the candidate reciept?  

> I would really appreciate if all information could be self contained.

It's unclear what you mean here.  We'll have the candidate reciept when fetching data, no?

We'd communicate something less than the full candidate reciept for off-chain messaging I guess.

> We don't need rotation for this: First core, first x validators are systemic chunk providers, next core , next x validators, ....

We absolutely require rotation of backing groups.   I'd give systemic chunks a different rotation, but not sure that matters.

We do want cores loaded as equally as possible of course.  We've at least two dimensions here, space and cpu time, so we'll have trouble even defining equally correctly, much less achieving it while we've this secondary market. 

> This can happen, but could be mitigated by incorporating in rewards: You reward people in proportion to the size of the chunks they served.  Should be possible to design this in a way, so there is no meaningful advantage in trying to manipulate for which cores your validators (or others) are responsible for.

Rewards are complex, if not chaotic.  Avoid complicating rewards unecessarily.

We've never discussed availability rewards being proportional to size.  Yes, proportional sounds interesting & nice.  Yet, rewards involve imprecissions or inequalities already, so proportional sounds like over commitment.  I'm unsure how proportional-to-size fits into tit-for-tat, for example. 

Rewards could acquire new bizarre constriants in future, via politics or competition, ala Cardano proves some Nash equilibrium, or craziness like cross-chain MEV, or stupidity like quadratic voting.  Also, rewards have bizarre conditionals because validators have different situations. 

Rewards solve major problems that overarch the whole project.  Rewards do not fine tune small problems. 

> Also in case this ever proves to be a real problem still, 

Would we notice?

> we could adjust the scheduler to move parachains between cores all the time, which would then still give us self contained systemic recovery, with evened out load. 

Yes, but isn't this much more complex than what we're discussing here?

> That is as bad as it gets

Yes, we're maybe polishing rocks here.

Reply:
Anyways..

Are you worried about slot numbers in off-chain messaging?

Or is there some simpler reason why slot numbers pose a problem?

Reply:
> It's unclear what you mean here. We'll have the candidate reciept when fetching data, no?

Yes, and we'd like availability-recovery to be possible when having nothing more than the candidate receipt. The problem is that the receipt does not contain any info about the slot or block number.
For the moment, the core index isn't there either, but the plan is to add it.

Reply:
I see, thank you.  We could only use some historical slot, not the slot where the candidate recipet gets backed.

All these options give some control of course, even the slot where backed, so relay parent hash works better than slots, due to simplicity. 

As I understand it @eskimor suggests our map depend upon only num_cores, num_validators, and core index, so each core has their systemic validators fixed for the session.  It avoids bias except through selecting your core.  I'd wager it worsens user experence, but only slightly. 

We do however rotate backing groups and backers provide systemic chunks too, hence the slightly above.  How do we determin the backers from the candidate recipet?  Just because they signed the candidate recipet? 

It's likely fine either way.  We'll anyways have systemic reconstruction if each systemic chunk can be fetched from some backers or its one availability provider. 

Actually who determines core index?  We'd ideas where this occurs after the candidate reciept.  We'd enable these if the map depends upon only num_cores, num_validators, relay parent, and paraid.

Reply:
> All these options give some control of course, even the slot where backed, so relay parent hash works better than any slot choice, due to simplicity.

As @eskimor said, "using the relay parent of the candidate is at the very least not ideal because with async backing, the used relay parent could vary from candidate to candidate in the same block, which means that the equal load distribution might end up not being that equal after all."

> As I understand it @eskimor suggests our map depend upon only num_cores, num_validators, and core index, so each core has their systemic validators fixed for the session. It avoids bias except through selecting your core. I'd wager it worsens user experence, but only slightly.

yes, that's the suggestion AFAIU.

> How do we determin the backers from the candidate recipet? Just because they signed the candidate recipet?

No, we only use the backers currently during approval-voting, when we have access to the relay block and we can query the session's validator groups from the runtime. For complexity reasons, we don't even verify that validators in the group all backed the candidate. We just assume that to be true (in practice, it's mostly true).

> Actually who determines core index? We'd ideas where this occurs after the candidate reciept. We'd enable these if the map depends upon only num_cores, num_validators, relay parent, and paraid.

Currently, for the existing lease parachains, there's a fixed assignment between the paraid and the core index. @eskimor you know more here

Reply:
Let's just go with only the core index, from the discussion I don't see any real problems with that and once we have the core index in the candidate receipt we are golden from the simplicity/robustness perspective.

Reply:
sounds good, I'll update the document